### PR TITLE
Use sorted set to limit with moving window

### PR DESCRIPTION
**Pull request by promag**

### CODE REVIEW

**Comment by noamshemesh** [comment]
I don't want to embed cpp here, is there any other option?  BTW the CI failed for node 4 and 5 (I will change to 6 soon)

**Comment by promag** [comment]
I've pushed a commit to use `process.hrtime()` instead of `microtime`. The idea is to low the chance of same score elements, ie, `zadd` has the highest chance to succeed.

File: index.js [hunk 1]
```diff
@@ -47,87 +47,38 @@ Limiter.prototype.inspect = function () {
...
-        if (isFirstReplyNull(res)) return mget();
-        n = n - 1;
-        done();
-      });
-  }
-
-  function mget() {
-    db.watch([count], function (err) {
+  var now = Date.now() * 1000 + Math.floor(process.hrtime()[1] / 1000) % 1000;
+  var start = now - duration * 1000;
```
Comment (promag) [inline index.js]: Should be `(duration + 1)` to take into account jitter from `hrtime` microseconds.

File: index.js [hunk 2]
```diff
@@ -47,87 +47,38 @@ Limiter.prototype.inspect = function () {
...
-        if (err) return fn(err);
-        if (isFirstReplyNull(res)) return mget();
-        n = n - 1;
-        done();
-      });
-  }
-
-  function mget() {
-    db.watch([count], function (err) {
+  var now = Date.now() * 1000 + Math.floor(process.hrtime()[1] / 1000) % 1000;
```
Comment (noamshemesh) [inline index.js]: I'm not sure using `process.hrtime` is the best approach either. By definition it related to arbitrary time in the past.

File: index.js (hunk 1 above)
Comment (noamshemesh) [inline index.js]: That's because `hrtime` is not stable enough for us :-\

**Comment by promag** [comment]
@noamshemesh sorry for the delay.

Current implementation problems:
- `MULTI` is not rollback'ed if the `WATCH`'ed key expires;
- Each key (count, limit and reset) have different expire timestamps and `MGET` can return some keys expired;
- `MULTI` is not rollback'ed if `SET NX` fails, so one key can be set while the others remain the same;

Given the above, and following the code, it can enter a loop `create -> mget -> create -> mget -> ...`, which results in a huge performance problem. This has been observed.

The way I see, the current implementation lacks for atomicity and it is unstable in a high concurrent system.

BTW, as stated [here](https://github.com/tj/node-ratelimiter/pull/30#issuecomment-277668948), tests fail randomly because of the above reasons.

The proposed solution doesn't suffer the above problems. It is only a bit more `redis` demanding
but much more concurrent friendly.

**Comment by promag** [comment]
@noamshemesh forgot to explain the need for microtime. `ZADD key score value` will add `value` to the sorted set and `score` must be unique. With `score` as milliseconds, even running the tests
causes `ZADD` to fail.

BTW, pushed a commit with an alternative solution regarding the microtime.

**Comment by promag** [comment]
Ping @noamshemesh.

**Comment by noamshemesh** [comment]
Sorry, crazy days, will take a look on the next weekend

**Comment by noamshemesh** [comment]
Looks good, thanks.
Please resolve the conflicts (it seems like you already fixed the previous PR, so just make sure you can ignore them), squash and I'll merge.

**Comment by promag** [comment]
@noamshemesh done.

BTW, if it would be awesome if you could create a release.

**Comment by noamshemesh** [comment]
Sure, I think I'll release a major version. what do you think?

**Comment by promag** [comment]
LGTM

**Comment by noamshemesh** [comment]
Done. Thanks a lot!

**Pull request merged**